hudi-common/src/main/java/org/apache/hudi/common/config/HoodieCommonConfig.java

```diff
@@ -38,7 +38,7 @@ public class HoodieCommonConfig extends HoodieConfig {
 
   public static final ConfigProperty<Boolean> RECONCILE_SCHEMA = ConfigProperty
       .key("hoodie.datasource.write.reconcile.schema")
-      .defaultValue(false)
+      .defaultValue(true)
```

Comment:
Initially was not in favor of this change, but now thinking about it a little more and especially in the light of https://github.com/apache/hudi/pull/6358, i think this is the right thing to do: for ex, after #6358, we'd be allowing to go writes, which might have columns dropped in the new batch. Now, there are 2 scenarios based on whether the reconciliation is enabled or not:

1. If reconciliation is _enabled_: we will be favoring table's schema and use it as a _writer-schema_. So in that case we will rewrite the incoming batch into the table's schema before applying it to the table.

2. If reconciliation is _disabled_: we will be favoring incoming batch's schema and use it as a _writer-schema_. In this case, for ex, for COW, we will be reading the table in its existing schema, but the new base files will be written in the writer's schema (ie w/ the column dropped)

Both of these approaches are legitimate and could be preferred in different circumstances. What's important here for us is to pick the right default setting that would minimize the _surprise effect_. 

Having reflected on this for some time now i think, that enabling reconciliation by default makes more sense as it protects table's schema from accidental mishaps in the incoming batches. And if somebody prefers the flow #2 the could easily opt-in for it by simply disabling the reconciliation.

WDYT?

Reply:
I agree. That was precisely the intention behind flipping this default.

Reply:
On the other hand, in its current form reconciliation doesn't allow to evolve the schema (unless comprehensive Schema Evolution is enabled) since it'll be essentially just favoring the table's schema always (there's no way for you to add new column for ex, other than switching off reconciliation)

Reply:
Is the difference mainly around the case of dropping a column?

Reply:
I’m thinking whether we should decouple the handling of drop column separately instead of turning on “schema reconciliation” by default, e.g., we should still allow new columns to be added instead of dropping them to favor table’s schema by default, while properly handling the column drop (maybe a different config?).

Reply:
I'll add my 5 cents as a hudi user :)
afaik schema reconciliation was meant to apply latest table schema on the incoming batch, but then if new batch contained new columns and some columns were missing at the same time then its behaviour was that new column was added but missing ones were dropped (at least on read, but physically still existing in files).

I feel like in Hudi we could have mergeSchema option for both df write and sql merge into (currently target table schema is applied) as in delta and now iceberg or in parquet datasource in spark, which would behave same as if reconciliation and schema evolution were enabled now. Then reconcile schema could behave differently.

When we ingest data the team producing it not always inform me about the changes and it would be nice to have a mechanism that can handle this. Currently most hudi users I know just create uber schema and apply it to df before write, sometimes it's hard to because of how the org we work for functions.

for some context:
#5899 - for mergeSchame in MERGE INTO statement
#5873 and #5452 - issues with reconcile schema

Reply:
@yihua it's more of a discussion what should be the default behavior:

 - Should we (by default) favor existing table's schema a SoT and rewrite incoming batch into it (unless Schema Evolution is enabled, in that case we will try to evolve the schema)
 - Should we (by default) favor incoming batch's schema as the schema we want table to be rewritten in

I still think that the #1 is a safer option as a default (optimizing for least amount of surprise to the user)

Reply:
@kazdy i think we just need to clearly disambiguate our configuration to make sure users can clearly understand what they can achieve and how (see my previous comment https://github.com/apache/hudi/pull/6196#discussion_r961984500): what you're describing could be achieved today enabling Reconciliation and Schema Evolution.

 -

Reply:
@alexeykudinkin afaik Schema Evolution config is there because it's an experimental feature and soon it will become GA? Then this config should be enabled by default or deprecated, will this logic hold then? I feel like hudi config is already very broad and therefore a bit hard to grasp and users would appreciate if it was one switch instead of a combination of two

Reply:
@kazdy correct, when Schema Evolution will become GA (cc @xiarixiaoyao) we will be flipping it to be on by default

Reply:
@alexeykudinkin @kazdy , now schema evolution cannot read by hive and presto, but we aready has pr to support that
https://github.com/apache/hudi/pull/6989
https://github.com/prestodb/presto/pull/18557
https://github.com/apache/hudi/pull/7045

once those pr merged, i think it will be ok.